### PR TITLE
feat: native board

### PR DESCRIPTION
**Pull request by kaspar030**
# Description

This PR provides an initial implementation of an infini-core scheduler, and a native board.

Try `laze -Cexamples/hello-world build -b native run`.

The scheduler implementation basically just removes the runqueue, and priorities are ignored. Each thread gets an `std::thread::Thread`.

The most tricky thing, (apart from enabling std for just some parts of the system), was to keep the isr based stopping of threads, where a call to `schedule()` pends an ISR, which triggers right at the end of a critical section. If that ISR de-schedules a thread into a non-Running state, that thread needs to stop. All our IPC logic depends on this behavior.
As std mutexes can only be unlocked from the thread that locked them, I used [`atomic-wait`](https://github.com/m-ou-se/atomic-wait) futexes, and provide a wrapper `critical-section` implementation that waits for a futex on `release()` if a thread is not supposed to continue. Runqueue operations are replaced a simple atomic stating whether a thread is "Running" or not, and the futex waits based on that.

[`atomic-wait`](https://github.com/m-ou-se/atomic-wait) hasn't seen much activity lately, but was written by Mara Bos herself, and seems to work well so far.

## Issues/PRs references

Depends on #1131.

~~Depends on #634 for output.~~

## Open Questions

- [ ] no peripherals so far, GPIO using tests/examples compile-fail
- [ ] networking
- [x] embedded-test

## Change checklist

- [ ] I have cleaned up my commit history and squashed fixup commits.
- [ ] I have followed the [Coding Conventions](https://ariel-os.github.io/ariel-os/dev/docs/book/coding-conventions.html).
- [ ] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by chrysn** [comment]
Awesome – looking forward to trying this out after the holidays!

Until we have a better place to spool ideas on peripherals, here's something I'd like to deposit (likely as a note-to-self): It would be highly convenient to have a BLE back-end that on the host side emulates a BLE adapter within radio reach of the device. That would speed up development a lot, both on the device side and for the host applications that interact with the device. (May even work in a WASM port of native that can run in the browser – with both the host webapp and a native board running in the same browser tab).

**Comment by chrysn** [comment]
This will need an entry in the `install-toolchain` target, but apart from that: The log example and hello-world-example work. Sure, peripherals will be needed, but that'll be added one by one.

Is this in a state where a code-level review can start?

(There's the aspect of documentation too, and we [gathered](https://matrix.to/#/!cINbgbHLemsitmThSh:matrix.org/$FFoig9i5mYt1yVQqoGl7oQbzhPgiN5kc0SqeWOJLDdI?via=matrix.org&via=utwente.io&via=club1.fr) a bit of "we should document its limitations", but I think this can go into the main tree and mature before we even have to do much in terms of having it as a documented feature.)

**Comment by kaspar030** [comment]
> Is this in a state where a code-level review can start?

I was kinda holding off until I get embedded-test to work, which upstream now has [initial support](https://github.com/probe-rs/embedded-test/pull/58). I'm almost there. :/

File: src/ariel-os-hal/src/dummy/mod.rs [hunk 1]
```diff
@@ -13,6 +13,7 @@
     reason = "this dummy module mimics manufacturer-specific crates"
 )]
 
+#[cfg(not(context = "native"))]
```
Comment (kaspar030) [inline src/ariel-os-hal/src/dummy/mod.rs]: Just three special cases in the dummy module allows using it for native w/o warnings.
`ariel-os-native` is basically empty now. I kept it so we have the scaffolding to add impls there.

File: src/ariel-os-native/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  "std",
+] }
+embedded-hal-async = { workspace = true }
+paste = { workspace = true }
+portable-atomic = { workspace = true }
+ariel-os-debug = { workspace = true, features = ["std"] }
+ariel-os-embassy-common = { workspace = true }
+ariel-os-random = { workspace = true, optional = true }
+
+[features]
```
Comment (kaspar030) [inline src/ariel-os-native/Cargo.toml:27]: These don't have any implementation behind, and are filtered by laze before they end up here. I kept them as scaffolding.

File: src/ariel-os-threads/src/sync/mutex.rs [hunk 3]
```diff
@@ -85,7 +85,7 @@ impl<T> Mutex<T> {
     /// # Panics
     ///
     /// Panics if called outside of a thread context.
-    pub fn lock(&self) -> MutexGuard<T> {
+    pub fn lock(&self) -> MutexGuard<'_, T> {
```
Comment (kaspar030) [inline src/ariel-os-threads/src/sync/mutex.rs]: These were suggested by clippy:
```
error: lifetime flowing from input to output with different syntax can be confusing                                                                            
  --> src/ariel-os-threads/src/sync/mutex.rs:88:17                                                                                                             
   |                                                                                                                                                           
88 |     pub fn lock(&self) -> MutexGuard<T> {                                                                                                                 
   |                 ^^^^^     ------------- the lifetime gets resolved as `'_`                                                                                
   |                 |                                                                                                                                         
   |                 this lifetime flows to the output                                                                                                         
   |                                                                                                                                                           
   = note: `-D mismatched-lifetime-syntaxes` implied by `-D warnings`                                                                                          
   = help: to override `-D warnings` add `#[allow(mismatched_lifetime_syntaxes)]`                                                                              
help: one option is to remove the lifetime for references and use the anonymous lifetime for paths                                                             
   |                                                                                                                                                           
88 |     pub fn lock(&self) -> MutexGuard<'_, T> {                                                                                                             
   |                                      +++

File: src/ariel-os-hal/src/dummy/mod.rs (hunk 1 above)
Comment (kaspar030) [reply src/ariel-os-hal/src/dummy/mod.rs]: Fixed in [`293b488` (#647)](https://github.com/ariel-os/ariel-os/pull/647/commits/293b48837d9554baa110331ab30d28dc9eb8758e).

File: src/ariel-os-threads/src/arch/native.rs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        pub static ID: Cell<Option<ThreadId>> = Cell::new(None);
+    }
+}
+
+impl Arch for Cpu {
+    type ThreadData = ThreadData;
+    const DEFAULT_THREAD_DATA: Self::ThreadData = ThreadData::new();
+
+    fn setup_stack(thread: &mut Thread, _stack: &mut [u8], func: usize, arg: usize) {
+        let func: fn(arg: usize) = unsafe { core::mem::transmute(func) };
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native.rs]: That transmute is possibly unsound, the docs say you need to [cast to a raw pointer before transmuting](https://doc.rust-lang.org/stable/core/primitive.fn.html#casting-to-and-from-integers). The `setup_stack()` method would also really need to be marked `unsafe` (with a proper `# Safety` doc-comment). But really even casting function pointers to a single `usize` is still [ABI dependent](https://doc.rust-lang.org/stable/core/primitive.fn.html#abi) and I would *really* prefer we first finally refactor the `thread::create_raw()` function to take a proper function pointer instead of [confusing addresses with pointers](https://doc.rust-lang.org/stable/core/ptr/index.html#provenance).

File: src/ariel-os-threads/src/arch/native/critical_section.rs
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native/critical_section.rs:1]: Seems to be mostly copied from [`rust-embedded/critical-section/src/std.rs`](https://github.com/rust-embedded/critical-section/blob/cebd3d76cc5237d77abd0e03c090577fad1c689f/src/std.rs)

File: src/ariel-os-threads/src/arch/native.rs (hunk 4 above)
Comment (kaspar030) [reply src/ariel-os-threads/src/arch/native.rs]: agreed. "works for now". Let's tackle that here: https://github.com/ariel-os/ariel-os/pull/1131

File: src/ariel-os-threads/src/sync/mutex.rs (hunk 3 above)
Comment (ROMemories) [reply src/ariel-os-threads/src/sync/mutex.rs]: Should be addressed by #1133.

File: src/ariel-os-native/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+] }
+embedded-hal-async = { workspace = true }
+paste = { workspace = true }
+portable-atomic = { workspace = true }
+ariel-os-debug = { workspace = true, features = ["std"] }
+ariel-os-embassy-common = { workspace = true }
+ariel-os-random = { workspace = true, optional = true }
+rand = { workspace = true, default-features = false, optional = true, features = [
+  "getrandom",
+] }
```
Comment (ROMemories) [inline src/ariel-os-native/Cargo.toml]: Seems like half these dependencies are unused, could you do some cleanup?

File: src/ariel-os-native/src/hwrng.rs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+/// Constructs the hardware random number generator (RNG) for the nRF family.
+///
```
Comment (ROMemories) [inline src/ariel-os-native/src/hwrng.rs]: Copy-paste leftover:

```suggestion
/// Constructs the hardware random number generator (RNG).
```

File: src/ariel-os-native/src/hwrng.rs [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+pub fn construct_rng(_peripherals: &mut crate::OptionalPeripherals) {
+    cfg_if::cfg_if! {
+        if #[cfg(any(context = "native"))] {
```
Comment (ROMemories) [inline src/ariel-os-native/src/hwrng.rs]:
```suggestion
        if #[cfg(context = "native")] {
```

File: src/ariel-os-runqueue/src/runqueue.rs [hunk 8]
```diff
@@ -12,6 +12,12 @@ const USIZE_BITS: usize = mem::size_of::<usize>() * 8;
 #[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub struct RunqueueId(u8);
 
+impl core::fmt::Display for RunqueueId {
+    fn fmt(&self, formatter: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+        core::fmt::Debug::fmt(self, formatter)
+    }
+}
```
Comment (ROMemories) [inline src/ariel-os-runqueue/src/runqueue.rs]: Do we really need these in this PR?

**Review by ROMemories — COMMENTED** [review]
Just a partial review for now, no need to start acting on my comments. Will do another pass later.

File: src/ariel-os-threads/src/arch/native.rs (hunk 4 above)
Comment (kaspar030) [reply src/ariel-os-threads/src/arch/native.rs]: This is now depending on #1131.

File: src/ariel-os-debug/src/lib.rs [hunk 9]
```diff
@@ -43,6 +43,9 @@ pub fn exit(code: ExitCode) {
     #[cfg(feature = "semihosting")]
     semihosting::process::exit(code.to_semihosting_code());
 
+    #[cfg(feature = "std")]
+    std::process::exit(code.to_semihosting_code());
```
Comment (ROMemories) [inline src/ariel-os-debug/src/lib.rs]: The exit code is platform-specific, the semihosting code can't just be reused like that.

File: src/ariel-os-debug/src/lib.rs [hunk 10]
```diff
@@ -195,7 +209,7 @@ pub use backend::*;
 #[doc(hidden)]
 #[cfg(feature = "log")]
 mod logger {
```
Comment (ROMemories) [inline src/ariel-os-debug/src/lib.rs]: Seems like changes to this module are bugfixes that could be extracted into a separate PR.

File: laze-project.yml [hunk 11]
```diff
@@ -1822,6 +1870,23 @@ builders:
...
+      - has_hwrng
+      - sw/benchmark
+    disables:
+      - semihosting
+      - defmt
+      - probe-rs
+    env:
+      RUSTC_TARGET: x86_64-unknown-linux-gnu
+      CARGO_TARGET_PREFIX: CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU
+      CARGO_RUNNER: '"sh -c"'
```
Comment (ROMemories) [inline laze-project.yml]: I think most of this should be a laze context (like MCUs) and we should have a separate builder (maybe also named `native`). That would also us to simulate whole boards when we have OoT board definitions.

File: src/ariel-os-threads/src/arch/native/critical_section.rs
Comment (ROMemories) [reply src/ariel-os-threads/src/arch/native/critical_section.rs:1]: Could you split this into two files: the `std` coming from upstream, and the Ariel OS one, so that it's easier to diff against the upstream one in the future?

File: src/ariel-os-threads/src/arch/native/critical_section.rs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+use std::sync::{Mutex, MutexGuard};
+
+static GLOBAL_MUTEX: Mutex<()> = Mutex::new(());
+
+// This is initialized if a thread has acquired the CS, uninitialized otherwise.
+static mut GLOBAL_GUARD: MaybeUninit<MutexGuard<'static, ()>> = MaybeUninit::uninit();
+
+std::thread_local!(static IS_LOCKED: Cell<bool> = Cell::new(false));
+
+struct StdCriticalSection;
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native/critical_section.rs]: Seems to me we could use the upstream one if (a) it was public and (b) didn't [set itself as the one critical section](https://github.com/rust-embedded/critical-section/blob/cebd3d76cc5237d77abd0e03c090577fad1c689f/src/std.rs#L13)?

File: src/ariel-os-threads/src/arch/native/critical_section.rs [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+struct ArielCriticalSection;
+unsafe impl critical_section::Impl for ArielCriticalSection {
+    unsafe fn acquire() -> bool {
+        unsafe { StdCriticalSection::acquire() }
+    }
+
+    unsafe fn release(nested_cs: bool) {
+        if !nested_cs {
+            let thread_id = crate::ThreadData::ID.get();
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native/critical_section.rs:52]: Seems like the ordering of this statement is important, could you add a comment explaining the relationship with the next statement?

File: laze-project.yml [hunk 14]
```diff
@@ -1526,22 +1539,57 @@ modules:
...
+    context:
+      # This overrides ariel-os::embedded-test-extra
+      - native
+    env:
+      global:
+        FEATURES:
+          - embedded-test/std
+        RUSTFLAGS:
+          - -Clink-arg=-Wl,-znostart-stop-gc
+          - -Clink-arg=-Wl,-static
```
Comment (ROMemories) [inline laze-project.yml]: Can you explain what these do/why they're needed?

File: laze-project.yml [hunk 15]
```diff
@@ -1526,22 +1539,57 @@ modules:
...
+    selects:
+      - doc-only
+
+  - name: embedded-test-runner-std
+    context: native
+    provides_unique:
+      - embedded-test-runner
+    env:
+      global:
+        CARGO_RUNNER: 'embedded-test-std-runner'
```
Comment (ROMemories) [inline laze-project.yml:1589]: I've found [this](https://github.com/kaspar030/embedded-test-std-runner), but not sure I understand how that's articulated.

File: src/ariel-os-hal/src/lib.rs [hunk 16]
```diff
@@ -30,7 +30,18 @@ pub mod define_peripherals;
...
+        pub use ariel_os_native::*;
+        #[path="dummy/gpio.rs"]
+        pub mod gpio;
+        #[path="dummy/identity.rs"]
+        pub mod identity;
+        #[path="dummy/peripheral.rs"]
+        pub mod peripheral;
+        pub fn init() -> OptionalPeripherals {
+            OptionalPeripherals {}
+        }
```
Comment (ROMemories) [inline src/ariel-os-hal/src/lib.rs]: Could you extract the commit that splits these dummy modules into a separate PR?

Comment (ROMemories) [reply src/ariel-os-hal/src/lib.rs]: It seems to me that these could then be more simply rewritten with something along the lines of:

```rust
mod dummy;

pub use dummy::{gpio, identity, peripheral};
```

File: laze-project.yml (hunk 14 above)
Comment (kaspar030) [reply laze-project.yml]: they actually aren't, I dropped them.

File: laze-project.yml (hunk 11 above)
Comment (kaspar030) [reply laze-project.yml]: interesting and good idea. builders share the namespace with contexts, so we'd need a different name.
Maybe tackle later? and, it is possible to have another board with native as parent.

File: src/ariel-os-debug/src/lib.rs (hunk 9 above)
Comment (kaspar030) [reply src/ariel-os-debug/src/lib.rs]: copy&pasted to `to_std_code()`

File: src/ariel-os-native/Cargo.toml (hunk 5 above)
Comment (kaspar030) [reply src/ariel-os-native/Cargo.toml]: I dropped paste and portable-atomic, I don't think native will ever need them. I kept the ones I think it will use (defmt, embassy-e-h, embedded-hal-async,ariel-os-embassy-common, ariel-os-debug)

File: src/ariel-os-runqueue/src/runqueue.rs (hunk 8 above)
Comment (kaspar030) [reply src/ariel-os-runqueue/src/runqueue.rs]: no, dropped and using `{:?}` now in the only user in this PR.

File: src/ariel-os-threads/src/arch/native/critical_section.rs (hunk 12 above)
Comment (kaspar030) [reply src/ariel-os-threads/src/arch/native/critical_section.rs]: Yeah. They might accept a PR.

File: src/ariel-os-threads/src/arch/native/critical_section.rs (hunk 13 above)
Comment (kaspar030) [reply src/ariel-os-threads/src/arch/native/critical_section.rs:52]: I've added some module level docs, and another comment here.

File: laze-project.yml (hunk 15 above)
Comment (kaspar030) [reply laze-project.yml:1589]: hm. how do we articulate this? that test runner needs to be installed (via `cargo install embedded-test-std-runner`). I've added a comment to the laze file.

File: src/ariel-os-hal/src/lib.rs (hunk 16 above)
Comment (kaspar030) [reply src/ariel-os-hal/src/lib.rs]: That would require re-adding the `cfg(context = "native")` that I had removed in the first place

Comment (ROMemories) [reply src/ariel-os-hal/src/lib.rs]: > That would require re-adding the `cfg(context = "native")` that I had removed in the first place

Not sure I get what you mean. Using my suggestion and adding
```rust
pub struct SWI;
```

to `ariel-os-native/src/lib.rs` seems to be building successfully.

Comment (kaspar030) [reply src/ariel-os-hal/src/lib.rs]: yes, done.

File: src/ariel-os-debug/src/lib.rs (hunk 10 above)
Comment (kaspar030) [reply src/ariel-os-debug/src/lib.rs]: split out the log comparison fix: https://github.com/ariel-os/ariel-os/pull/1152

Comment (kaspar030) [reply src/ariel-os-debug/src/lib.rs]: let's piggy-back the other two changes.

**Comment by ROMemories** [comment]
Seems like the HWRNG implementation is orthogonal to the rest, could you also move it to a follow-up, so this PR is even smaller and easier for me to review?

**Comment by kaspar030** [comment]
> Seems like the HWRNG implementation is orthogonal to the rest, could you also move it to a follow-up, so this PR is even smaller and easier for me to review?

Removed the hwrng impl from this PR, will open another once this is in!

File: ariel-os-cargo.toml [hunk 17]
```diff
@@ -25,6 +25,7 @@ embassy-nrf = { git = "https://github.com/ariel-os/embassy", branch = "embassy-n
 embassy-net = { git = "https://github.com/ariel-os/embassy", branch = "embassy-net-v0.6.0+ariel-os" }
 embassy-rp = { git = "https://github.com/ariel-os/embassy", branch = "embassy-rp-v0.4.0+ariel-os+trng-panic-fix" }
 embassy-stm32 = { git = "https://github.com/ariel-os/embassy", branch = "embassy-stm32-v0.2.0+ariel-os" }
+embassy-time = { git = "https://github.com/ariel-os/embassy", branch = "embassy-time-v0.4.0+ariel-os" }
```
Comment (ROMemories) [inline ariel-os-cargo.toml:28]: For the record: currently all this does is drop the critical section impl auto-enabled by the `std` feature, so that it does not conflict with the implementation provided in this PR.

File: laze-project.yml [hunk 18]
```diff
@@ -1822,6 +1868,22 @@ builders:
...
+    provides:
+      - sw/benchmark
+    disables:
+      - semihosting
+      - defmt
+      - probe-rs
+    env:
+      RUSTC_TARGET: x86_64-unknown-linux-gnu
+      CARGO_TARGET_PREFIX: CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU
+      CARGO_RUNNER: '"sh -c"'
```
Comment (ROMemories) [inline laze-project.yml:1885]: Why do we need the double quotes?

File: laze-project.yml (hunk 15 above)
Comment (ROMemories) [reply laze-project.yml:1589]: That'll definitely need to go into the docs when we have some (later on).

File: src/ariel-os-debug/src/lib.rs [hunk 19]
```diff
@@ -1,6 +1,6 @@
 //! Provides debug interface facilities.
 
-#![cfg_attr(not(test), no_std)]
+#![cfg_attr(not(any(test, context = "native")), no_std)]
```
Comment (ROMemories) [inline src/ariel-os-debug/src/lib.rs]: Shouldn't that use the newly introduced Cargo feature instead?

```suggestion
#![cfg_attr(not(any(test, feature = "std")), no_std)]
```

File: laze-project.yml [hunk 20]
```diff
@@ -1526,22 +1539,55 @@ modules:
...
         cmd:
           - ${CARGO_ENV} ${CARGO} ${CARGO_TOOLCHAIN} ${CARGO_ARGS} test ${FEATURES} $@
         build: false
-
     env:
       global:
         RUSTFLAGS:
           - -Clink-arg=-Tembedded-test.x
 
+  - name: embedded-test-extra
```
Comment (ROMemories) [inline laze-project.yml:1558]: Not a huge fan of the name of these modules, but considering they're internal we can just renamed them later.

File: src/ariel-os-hal/src/lib.rs [hunk 21]
```diff
@@ -30,7 +30,11 @@ pub mod define_peripherals;
 pub use define_peripherals::*;
 
 cfg_if::cfg_if! {
-    if #[cfg(context = "nrf")] {
+    if #[cfg(context = "native")] {
+        pub use ariel_os_native::*;
+        pub mod dummy;
+        pub use dummy::{gpio, identity, peripheral};
```
Comment (ROMemories) [inline src/ariel-os-hal/src/lib.rs]: No need to make the module `pub`, and I would really prefer we kept module *declarations* before `use`s:

```suggestion
        mod dummy;

        pub use ariel_os_native::*;
        pub use dummy::{gpio, identity, peripheral};
```

File: src/ariel-os-native/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+
+#![no_std]
+#![cfg_attr(nightly, feature(doc_auto_cfg))]
+
+pub struct OptionalPeripherals {}
+pub struct SWI {}
+
+pub fn init() -> OptionalPeripherals {
+    OptionalPeripherals {}
+}
```
Comment (ROMemories) [inline src/ariel-os-native/src/lib.rs]: I would prefer we kept the same order as in the dummy module, that makes comparing them easier:

```suggestion

pub fn init() -> OptionalPeripherals {
    OptionalPeripherals {}
}

pub struct SWI {}
```

File: src/ariel-os-power/src/lib.rs [hunk 23]
```diff
@@ -1,7 +1,7 @@
 //! Provides power management functionality.
 
 #![deny(missing_docs)]
-#![no_std]
+#![cfg_attr(not(context = "native"), no_std)]
```
Comment (ROMemories) [inline src/ariel-os-power/src/lib.rs:4]: I think this crate should get an `std` Cargo feature and use it here.

File: src/ariel-os-threads/src/arch/native.rs [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+
+impl ThreadData {
+    pub const fn new() -> Self {
+        Self { thread: None }
+    }
+}
+
+impl ThreadData {
+    thread_local! {
+        pub static ID: Cell<Option<ThreadId>> = Cell::new(None);
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native.rs]: Seems like `thread_local!` supports the following:

```suggestion
        pub static ID: Cell<Option<ThreadId>> = const { Cell::new(None) };
```

File: src/ariel-os-threads/src/arch/native.rs [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+
+    fn schedule() {}
+
+    fn wfi() {
+        unimplemented!()
+    }
+
+    fn set_running(thread_id: ThreadId) {
+        let n = usize::from(thread_id);
+        if THREAD_BLOCKS[n].swap(1, Ordering::Acquire) == 0 {
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native.rs]: Seems like both `0` and `1` are magic values of `THREAD_BLOCKS`; they would benefit from having constants defined for them (especially as these specific integers may well not be magic in other contexts).

File: src/ariel-os-threads/src/arch/native.rs [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+                    //
+                    // Whether this results in a function pointer that is safe to call:
+                    //
+                    // If `arg.is_some()`, `func` has been transmuted from `fn(T: Arguable + Send)`,
+                    // here we transmute it again into a function pointer with one `usize` argument.
+                    //
+                    // According to
+                    // https://doc.rust-lang.org/stable/std/primitive.fn.html#abi-compatibility,
+                    // this "function call behaves as if every argument was transmuted from the
+                    // type in the function pointer to the type at the function declaration".
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native.rs:58]: > **If the declared signature and the signature of the function pointer are ABI-compatible**, then the function call behaves as if every argument was [transmuted](https://doc.rust-lang.org/stable/std/mem/fn.transmute.html) from the type in the function pointer to the type at the function declaration, and the return value is [transmuted](https://doc.rust-lang.org/stable/std/mem/fn.transmute.html) from the type in the declaration to the type in the pointer.

Also:

> For two signatures to be considered ABI-compatible, they must use a compatible ABI string, must take the same number of arguments […]

`fn()` and `fn(T)` are of course not ABI-compatible

File: src/ariel-os-threads/src/arch/native.rs [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+    const DEFAULT_THREAD_DATA: Self::ThreadData = ThreadData::new();
+
+    fn setup_stack(thread: &mut Thread, _stack: &mut [u8], func: fn(), arg: Option<usize>) {
+        let thread_id = thread.tid;
+
+        let handle = std::thread::spawn(move || {
+            ThreadData::ID.with(|x| x.set(Some(thread_id)));
+            atomic_wait::wait(&THREAD_BLOCKS[usize::from(thread_id)], 0);
+
+            let res = std::panic::catch_unwind(|| {
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native.rs:45]: What's the purpose of this `catch_unwind` here?

File: src/ariel-os-threads/src/arch/native.rs [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+            });
+        });
+
+        thread.data.thread = Some(handle.thread().clone());
+    }
+
+    fn start_threading() {
+        loop {
+            SCHEDULER.with(|scheduler| {
+                for (n, thread) in scheduler.threads.iter().enumerate() {
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native.rs:85]: optional: I think you could `zip()` `THREAD_BLOCKS.iter()` to avoid indexing.

File: src/ariel-os-threads/src/lib.rs [hunk 29]
```diff
@@ -201,10 +218,11 @@ impl Scheduler {
         _core_affinity: Option<CoreAffinity>,
     ) -> Option<ThreadId> {
         let (thread, tid) = self.get_unused()?;
-        Cpu::setup_stack(thread, stack, func, arg);
         thread.prio = prio;
         thread.tid = tid;
         thread.state = ThreadState::Parked;
+        Cpu::setup_stack(thread, stack, func, arg);
```
Comment (ROMemories) [inline src/ariel-os-threads/src/lib.rs:227]: Is this a bug fix or is the difference only visible for native?

File: src/ariel-os-threads/src/arch/native/critical_section.rs [hunk 30]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    }
+}
+
+critical_section::set_impl!(ArielCriticalSection);
+
+struct ArielCriticalSection;
+unsafe impl critical_section::Impl for ArielCriticalSection {
+    unsafe fn acquire() -> bool {
+        unsafe { StdCriticalSection::acquire() }
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native/critical_section.rs:44]: This `unsafe` block should have a `SAFETY` comment.

File: src/ariel-os-threads/src/arch/native/critical_section.rs [hunk 31]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    unsafe fn release(nested_cs: bool) {
+        if !nested_cs {
+            let thread_id = crate::ThreadData::ID.get();
+            unsafe { StdCriticalSection::release(false) }
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native/critical_section.rs:54]: This `unsafe` block should have a `SAFETY` comment.

File: src/ariel-os-threads/src/arch/native/critical_section.rs [hunk 32]
```diff
@@ -0,0 +1,110 @@
+//! critical_section implementation for std & infinicore
+//!
+//! In Ariel OS, most IPC code expects a context switch that is triggered within a
+//! critical section to be deferred until the critical section ends, and then run
+//! in an ISR.
+//! This means, a thread might get scheduled away right after a critical section,
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native/critical_section.rs]:
```suggestion
//! This means a thread might get scheduled away right after a critical section,
```

File: src/ariel-os-threads/src/arch/native/critical_section.rs [hunk 33]
```diff
@@ -0,0 +1,110 @@
...
+//! maybe because it is not runnable anymore (but waiting for a lock, message or
+//! flag).
+//!
+//! On native with each Ariel thread running in its own host kernel thread, we
+//! need to emulate this behavior. So we wrap the regular `StdCriticalSection`.
+//! We keep an `AtomicLock` per thread. Instead of the regular runqueue for keeping
+//! track of runnable threads, the scheduler locks/unlocks those `AtomicLock`s
+//! to specify whether a thread is runnable, and the wrapped critical section, on `release()`
```
Comment (ROMemories) [inline src/ariel-os-threads/src/arch/native/critical_section.rs]: I'm having trouble parsing this line.

File: src/ariel-os-native/Cargo.toml [hunk 34]
```diff
@@ -0,0 +1,47 @@
+[package]
+name = "ariel-os-native"
+version = "0.1.0"
```
Comment (ROMemories) [inline src/ariel-os-native/Cargo.toml]: Seems like it would be easier if we kept it in sync with the other crates:

```suggestion
version = "0.2.0"
```

File: src/ariel-os-power/src/lib.rs [hunk 35]
```diff
@@ -16,6 +16,8 @@ pub fn reboot() -> ! {
             esp_hal::reset::software_reset();
             #[expect(clippy::empty_loop, reason = "software_reset() should eventually never return")]
             loop {}
+        } else if #[cfg(context = "native")] {
+            std::process::exit(0)
```
Comment (ROMemories) [inline src/ariel-os-power/src/lib.rs:20]: This doesn't actually reboot; maybe rephrase the documentation in a way that allows that behavior, without actually referring to native (as I don't we want to document it just yet?) and without making that a breaking change?

File: ariel-os-cargo.toml (hunk 17 above)
Comment (kaspar030) [reply ariel-os-cargo.toml:28]: yes, and that patch has been merged upstream

File: src/ariel-os-threads/src/arch/native.rs (hunk 26 above)
Comment (ROMemories) [reply src/ariel-os-threads/src/arch/native.rs:58]: After some OoB discussions and having another read at [the docs](https://doc.rust-lang.org/stable/std/primitive.fn.html#abi-compatibility), it seems that the ABI-compatibility requirements for function pointers and function signatures are actually different, and that `fn()` and `fn(T)` (as function pointers) are actually ABI-compatible.

File: laze-project.yml (hunk 18 above)
Comment (kaspar030) [reply laze-project.yml:1885]: because we need this to extrapolate to `CARGO_RUNNER="sh -c"` (with quotes). Otherwise the command line becomes `CARGO_RUNNER=sh -c cargo ...`.
using single and double quotes so the double quotes don't get eaten by yaml.

File: src/ariel-os-power/src/lib.rs (hunk 23 above)
Comment (kaspar030) [reply src/ariel-os-power/src/lib.rs:4]: I'm somewhat reluctant here. The file uses contexts to distinguish between targets later on.

File: src/ariel-os-threads/src/arch/native.rs (hunk 27 above)
Comment (kaspar030) [reply src/ariel-os-threads/src/arch/native.rs:45]: I'll add a comment.

`catch_unwind` catches a panicking thread and allows us to handle that. Without that, threads would panic but Ariel would happily continue (with that thread in some weird state). So to match other platforms' behavoir, abort on any thread panic.

File: src/ariel-os-threads/src/lib.rs (hunk 29 above)
Comment (kaspar030) [reply src/ariel-os-threads/src/lib.rs:227]: native's `setup_stack()` needs the `tid` field populated. I'll add a comment.

File: src/ariel-os-threads/src/arch/native/critical_section.rs (hunk 33 above)
Comment (kaspar030) [reply src/ariel-os-threads/src/arch/native/critical_section.rs]: reworded

File: src/ariel-os-debug/src/lib.rs (hunk 10 above)
Comment (kaspar030) [reply src/ariel-os-debug/src/lib.rs]: both merged in #1152

File: src/ariel-os-power/src/lib.rs (hunk 35 above)
Comment (kaspar030) [reply src/ariel-os-power/src/lib.rs:20]: Alternatively, print sth like "native: reboot called, exiting..." ?
IMO there is no concept of "exiting" on MCUs.
And I checked, there seems to be no portable way of re-execing a process, so no way of mimicking a reboot by the binary itself.
Another option would be to exit with a specific exit code and re-start the binary from the outside, maybe using a runner that is smarter than "sh -c".

File: src/ariel-os-threads/src/arch/native.rs (hunk 27 above)
Comment (kaspar030) [reply src/ariel-os-threads/src/arch/native.rs:45]: done

File: src/ariel-os-threads/src/arch/native.rs (hunk 25 above)
Comment (kaspar030) [reply src/ariel-os-threads/src/arch/native.rs]: I have renamed the array to `THREAD_RUNNABLE`. I'd rather not do defines like `const THREAD_IS_RUNNABLE: usize = 1` and `const THREAD_IS_NOT_RUNNABLE: usize = 0`, IMO that is just cruft and mental overhead.

File: laze-project.yml (hunk 18 above)
Comment (ROMemories) [reply laze-project.yml:1885]: So the double quotes are not for shell expansion? Usually I would prefer to stick to single quotes then (so in this case that would be using the double the double quotes for YAML, and single quotes for the shell); but fine to keep them here.

File: src/ariel-os-power/src/lib.rs (hunk 35 above)
Comment (ROMemories) [reply src/ariel-os-power/src/lib.rs:20]: It was a bit surprising at first not getting the same debug output, but I get what you say and we can reconsider later.

**Review by ROMemories — APPROVED** [review]
Successfully re-did the testing of my previous review, and the `tests/threading-lock` test is now disabled.

There is quite a few dead code warnings when compiling the examples, but I would rather we address them in a follow-up.
